runelite-client/src/main/java/net/runelite/client/plugins/mining/MiningPlugin.java

```diff
@@ -161,6 +203,68 @@ public void onWallObjectSpawned(WallObjectSpawned event)
 		}
 	}
 
+	@Subscribe
+	public void onMenuOptionClicked(MenuOptionClicked event)
+	{
+		if (event.getWidgetId() != WidgetInfo.INVENTORY.getId() || event.getMenuAction() == MenuAction.RUNELITE)
+		{
+			return;
+		}
+
+		ItemContainer inventoryItemContainer = client.getItemContainer(InventoryID.INVENTORY);
+
+		switch (event.getMenuOption().toLowerCase())
+		{
+			case FILL_OPTION:
+				updateAmountOfCoalInBag((int) Arrays.stream(inventoryItemContainer.getItems()).filter(i -> i.getId() == ItemID.COAL).count());
+				break;
+			case EMPTY_OPTION:
+				int amt = MAX_INVENTORY_SPACE - (int) Arrays.stream(inventoryItemContainer.getItems()).filter(i -> i.getId() != -1).count();
+				updateAmountOfCoalInBag(-amt);
+				break;
+		}
+	}
+
+	/**
+	 * Use onChatMessage when a player chooses the `Check` menu option on coal bags.
+	 *
+	 * @param event
+	 */
+	@Subscribe
+	public void onChatMessage(ChatMessage event)
+	{
+		String chatMsg = Text.removeTags(event.getMessage());
+		if (COAL_BAG_EMPTY_MESSAGE.matcher(chatMsg).find())
+		{
+			setAmountOfCoalInBag(0);
+		}
+		else if (COAL_BAG_ONE_MESSAGE.matcher(chatMsg).find())
+		{
+			setAmountOfCoalInBag(1);
+		}
+		else
+		{
+			Matcher matcher = COAL_BAG_AMOUNT_MESSAGE.matcher(chatMsg);
+			if (matcher.find())
+			{
+				updateAmountOfCoalInBag(Integer.parseInt(matcher.group(1)) - getAmountOfCoalInBag());
+			}
+		}
+	}
+
+	/**
+	 * Update the player's count of coal in their Coal Bag
+	 *
+	 * @param delta How much to add/subtract from the amount.
+	 *              Supply a negative number to subtract, or positive number to add.
+	 */
+	protected void updateAmountOfCoalInBag(int delta)
+	{
+		// check for upper/lower bounds of amount of coal in a bag
+		// 0 <= X <= 27
+		setAmountOfCoalInBag(Math.max(0, Math.min(FULL_COAL_BAG_AMOUNT, getAmountOfCoalInBag() + delta)));
```

Comment:
Ints::constrainToRange does not serve the same purpose as what I have here exactly. This will not update the value if the new value pushes the amount over the 27 item threshold - constraining to range updates to a new value each time, as found in testing.